pkg/controllers/multiclusterservice/endpointslice_dispatch_controller.go

```diff
@@ -0,0 +1,279 @@
+			continue
+		}
+
+		// It couldn't happen here
+		if len(work.Spec.Workload.Manifests) == 0 {
+			continue
+		}
+
+		// There should be only one manifest in the work, let's use the first one.
+		manifest := work.Spec.Workload.Manifests[0]
+		unstructuredObj := &unstructured.Unstructured{}
+		if err := unstructuredObj.UnmarshalJSON(manifest.Raw); err != nil {
+			klog.Errorf("Failed to unmarshal work manifest, error is: %v", err)
+			return err
+		}
+
+		endpointSlice := &discoveryv1.EndpointSlice{}
+		if err := helper.ConvertToTypedObject(unstructuredObj, endpointSlice); err != nil {
+			klog.Errorf("Failed to convert unstructured object to typed object, error is: %v", err)
+			return err
+		}
+
+		// Use this name to avoid naming conflicts and locate the EPS source cluster.
+		endpointSlice.Name = epsSourceCluster + "-" + endpointSlice.Name
+		clusterNamespace := names.GenerateExecutionSpaceName(clusterName)
+		endpointSlice.Labels = map[string]string{
+			discoveryv1.LabelServiceName:    mcs.Name,
+			workv1alpha1.WorkNamespaceLabel: clusterNamespace,
+			workv1alpha1.WorkNameLabel:      work.Name,
+			util.ManagedByKarmadaLabel:      util.ManagedByKarmadaLabelValue,
+			discoveryv1.LabelManagedBy:      util.EndpointSliceControllerLabelValue,
+		}
+		endpointSlice.Annotations = map[string]string{
+			util.EndpointSliceProvisionClusterAnnotation:    epsSourceCluster,
+			util.EndPointSliceProvisionGenerationAnnotation: strconv.FormatInt(endpointSlice.Generation, 10),
+		}
+
+		workMeta := metav1.ObjectMeta{
+			Name:       work.Name,
+			Namespace:  clusterNamespace,
+			Finalizers: []string{util.ExecutionControllerFinalizer},
+			Labels: map[string]string{
+				util.EndpointSliceWorkProvisionClusterLabel: epsSourceCluster,
+				util.ManagedByKarmadaLabel:                  util.ManagedByKarmadaLabelValue,
+			},
+		}
+		unstructuredEPS, err := helper.ToUnstructured(endpointSlice)
+		if err != nil {
+			klog.Errorf("Failed to convert typed object to unstructured object, error is: %v", err)
+			return err
+		}
+		if err := helper.CreateOrUpdateWork(c.Client, workMeta, unstructuredEPS); err != nil {
+			klog.Errorf("Failed to sync EndpointSlice %s/%s from %s to cluster %s:%v",
+				work.GetNamespace(), work.GetName(), epsSourceCluster, clusterName, err)
+			return err
+		}
+	}
+
+	if !controllerutil.ContainsFinalizer(work, util.MCSEndpointSliceDispatchControllerFinalizer) {
+		controllerutil.AddFinalizer(work, util.MCSEndpointSliceDispatchControllerFinalizer)
+		if err := c.Client.Update(ctx, work); err != nil {
+			klog.Errorf("Failed to add finalizer %s for work %s/%s, error: %v", util.MCSEndpointSliceDispatchControllerFinalizer, work.Namespace, work.Name, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *EndpointsliceDispatchController) cleanupEndpointSliceFromConsumerClusters(ctx context.Context, work *workv1alpha1.Work) error {
+	// TBD: There may be a better way without listing all works.
+	workList := &workv1alpha1.WorkList{}
```

Comment:
Can we filter work by such as LabelSelector?

Reply:
I have no idea how to define the label, let's optimize it lattter